charmhelpers/fetch/ubuntu.py

```diff
@@ -189,17 +190,185 @@
+
+    Note that one of the main purposes for this function is to avoid dependency
+    on the ``python-apt`` package.
+
+    The indicated package is a wrapper around the ``apt`` C++ library which is
+    tightly connected to the version of the distribution it is shipped on.
+    This in turn makes it incredibly hard to distribute as a wheel for a charm
+    that supports a large span of distro versions.
+
+    :param packages: Packages to get data from
+    :type packages: list[str]
+    :returns: Structured data about package, keys like ``apt-cache show``
+    :rtype: dict
+    :raises: subprocess.CalledProcessError
+    """
+    pkgs = {}
+    cmd = ['apt-cache', 'show', '--no-all-versions']
+    cmd.extend(packages)
+    if locale.getlocale() == (None, None):
+        # subprocess calls out to locale.getpreferredencoding(False) to
+        # determine encoding.  Workaround for Trusty where the
+        # environment appears to not be set up correctly.
+        locale.setlocale(locale.LC_ALL, 'en_US.UTF-8')
+    try:
+        output = subprocess.check_output(cmd,
+                                         stderr=subprocess.STDOUT,
+                                         universal_newlines=True)
+        previous = None
+        pkg = {}
+        for line in output.splitlines():
+            if not line:
+                if 'package' in pkg:
+                    pkgs.update({pkg['package']: pkg})
+                    pkg = {}
+                continue
+            if line.startswith(' '):
+                if previous and previous in pkg:
+                    pkg[previous] += os.linesep + line.lstrip()
+                continue
+            if ':' in line:
+                kv = line.split(':', 1)
+                key = kv[0].lower()
+                if key == 'n':
+                    continue
+                previous = key
+                pkg.update({key: kv[1].lstrip()})
+    except subprocess.CalledProcessError as cp:
+        # ``apt-cache`` returns 100 if none of the packages asked for exist in
+        # the apt cache.
+        if cp.returncode != 100:
+            raise
+    return pkgs
+
+
+def upstream_version(version):
+    """Extracts upstream version from a version string.
+
+    Note that one of the main purposes for this function is to avoid dependency
+    on the ``python-apt`` package.
+
+    The indicated package is a wrapper around the ``apt`` C++ library which is
+    tightly connected to the version of the distribution it is shipped on.
+    This in turn makes it incredibly hard to distribute as a wheel for a charm
+    that supports a large span of distro versions.
+
+    :param version: Version string
+    :type version: str
+    :returns: Upstream version
+    :rtype: str
+
+    Upstream reference: https://salsa.debian.org/apt-team/apt/blob/master/
+                                apt-pkg/deb/debversion.cc#L259
+    """
+    if version:
+        version = version[version.find(':') + 1:]
```

Comment:
Alternative:

```python
    version = version.split(':')[-1]
```

It will always pass, even if there is no ':' in the string (it returns the whole string).  if you are worried about more than one ':', use `split(':', 1)`

The `+ 1` makes it a bit more interesting.  Add a `[1:]` after the `[-1]` to strip off the first char.

Reply:
Thank you for the tip, have incorporated this in the function now.